src/picongpu/include/particles/ionization/byField/BSI/BSI.hpp

```diff
@@ -32,30 +45,147 @@ namespace particles
 namespace ionization
 {
 
-    /** \struct BSI
-     * \brief Barrier Suppression Ionization
+    /** \struct BSI_Impl
+     * \brief Barrier Suppression Ionization - Implementation
+     *
+     * \tparam T_DestSpecies electron species to be created
+     * \tparam T_SrcSpecies particle species that is ionized
      */
-    template<typename T_DestSpecies>
-    struct BSI
+    template<typename T_DestSpecies, typename T_SrcSpecies>
+    struct BSI_Impl
     {
 
-        /* define ionization ALGORITHM for ionization MODEL */
-        typedef particles::ionization::AlgorithmBSI IonizationAlgorithm;
-
-        /** Functor implementation
-         * \tparam T_SrcSpecies particle species to be ionized
-         * \tparam T_ParticleStorage \see picongpu/src/libPMacc/include/math/MapTuple.hpp
-         *
-         * \param currentStep current time step */
-        template<typename T_SrcSpecies,typename T_ParticleStorage>
-        void operator()(T_SrcSpecies& src, T_ParticleStorage& pst, const uint32_t currentStep)
-        {
-
-            /* alias for electron species to be created */
-            PMACC_AUTO(electrons,pst[typename MakeIdentifier<T_DestSpecies>::type()]);
-            /* call the ionization function */
-            src.ionize(electrons, currentStep);
-        }
+        typedef T_DestSpecies DestSpecies;
+        typedef T_SrcSpecies  SrcSpecies;
+
+        typedef typename SrcSpecies::FrameType FrameType;
+
+        /* specify field to particle interpolation scheme */
+        typedef typename PMacc::traits::Resolve<
+            typename GetFlagType<FrameType,interpolation<> >::type
+        >::type InterpolationScheme;
+
+        /* margins around the supercell for the interpolation of the field on the cells */
+        typedef typename GetMargin<InterpolationScheme>::LowerMargin LowerMargin;
+        typedef typename GetMargin<InterpolationScheme>::UpperMargin UpperMargin;
+
+        /* relevant area of a block */
+        typedef SuperCellDescription<
+            typename MappingDesc::SuperCellSize,
+            LowerMargin,
+            UpperMargin
+            > BlockArea;
+
+        BlockArea BlockDescription;
+
+        private:
+
+            /* define ionization ALGORITHM (calculation) for ionization MODEL */
+            typedef particles::ionization::AlgorithmBSI IonizationAlgorithm;
+
+            typedef MappingDesc::SuperCellSize TVec;
+            typedef FieldE::ValueType ValueType;
+            typedef FieldE::DataBoxType DataBoxType;
+            /* global memory EM-field device databoxes */
+            DataBoxType eBox;
```

Comment:
if I wrap it in a `PMACC_ALIGN()` and when I then pass the ionization model by reference to `const`, I get problems with the member functions of the ionization model:

*error: the object has type qualifiers that are not compatible with the member function*

This occurs when I call `frameIonizer.init( ... )` on device where I am caching the fields